pkg/lorry/engines/custom/manager.go

```diff
@@ -83,78 +98,184 @@ func NewManager(properties engines.Properties) (engines.DBManager, error) {
 		Transport: netTransport,
 	}
 
-	return mgr, nil
+	return nil
 }
 
-func (mgr *Manager) GetReplicaRole(ctx context.Context, cluster *dcs.Cluster) (string, error) {
-	if mgr.actionSvcPorts == nil {
-		return "", nil
+func (mgr *Manager) InitComponentDefintionActions() error {
+	actionJSON := viper.GetString(constant.KBEnvActionCommands)
+	if actionJSON != "" {
+		err := json.Unmarshal([]byte(actionJSON), &mgr.actionCommands)
+		if err != nil {
+			return err
+		}
 	}
+	return nil
+}
 
-	var (
-		lastOutput []byte
-		err        error
-	)
+// JoinCurrentMemberToCluster provides the following dedicated environment variables for the action:
+//
+// - KB_SERVICE_PORT: The port on which the DB service listens.
+// - KB_SERVICE_USER: The username used to access the DB service with sufficient privileges.
+// - KB_SERVICE_PASSWORD: The password of the user used to access the DB service .
+// - KB_PRIMARY_POD_FQDN: The FQDN of the original primary Pod before switchover.
+// - KB_NEW_MEMBER_POD_NAME: The name of the new member's Pod.
+func (mgr *Manager) JoinCurrentMemberToCluster(ctx context.Context, cluster *dcs.Cluster) error {
+	memberJoinCmd, ok := mgr.actionCommands[constant.MemberJoinAction]
+	if !ok || len(memberJoinCmd) == 0 {
+		return errors.New("member join command is empty!")
+	}
+	envs, err := util.GetGlobalSharedEnvs()
+	if err != nil {
+		return err
+	}
 
-	for _, port := range *mgr.actionSvcPorts {
-		u := fmt.Sprintf("http://127.0.0.1:%d/role?KB_RSM_LAST_STDOUT=%s", port, url.QueryEscape(string(lastOutput)))
-		lastOutput, err = mgr.callAction(ctx, u)
-		if err != nil {
-			return "", err
-		}
-		mgr.Logger.Info("action succeed", "url", u, "output", string(lastOutput))
-	}
-	finalOutput := strings.TrimSpace(string(lastOutput))
-
-	if perNodeRegx.MatchString(finalOutput) {
-		return finalOutput, nil
-	}
-
-	// csv format: term,podName,role
-	parseCSV := func(input string) (string, error) {
-		res := common.GlobalRoleSnapshot{}
-		lines := strings.Split(input, "\n")
-		for _, line := range lines {
-			fields := strings.Split(strings.TrimSpace(line), ",")
-			if len(fields) != 3 {
-				return "", err
-			}
-			res.Version = strings.TrimSpace(fields[0])
-			pair := common.PodRoleNamePair{
-				PodName:  strings.TrimSpace(fields[1]),
-				RoleName: strings.ToLower(strings.TrimSpace(fields[2])),
-			}
-			res.PodRoleNamePairs = append(res.PodRoleNamePairs, pair)
-		}
-		resByte, err := json.Marshal(res)
-		return string(resByte), err
+	if cluster.Leader == nil || cluster.Leader.Name == "" {
```

Comment:
This constraint is not needed for 'stateless' workloads, even for stateful workloads, the members in the statefulset are identical and have the same functions and roles. For example, for the nodes in Oracle RAC, the join operation may need the FQDN of a manager service, for the nodes for nginx behind a load balancer, the join operation only registers itself to a service, for the BE nodes for StarRocks,  the join operation needs to contact with the FE.

Reply:
It is better to goes such way: if there is a replication relation between roles(paxos or master-slave), the leader/primary/master/follower/... works as the traditional way, if they have NONE role, the developer also has the ability to write a 'Join' action and get executed, the 'Join' action body may be only a 'echo xxxx'. It is also very important for the developer to 'echo' something to the log to do further debugging & monitoring & alerting.

Reply:
It will be very very annoyed for a developer to write a huge body of 'Join' and gets no effect only because it has no leader role.....